src/devices/honeywell_cm921.c

```diff
@@ -0,0 +1,449 @@
+    *out = 0;
+    int bend = pos + 10;
+
+    while (pos < bend) {
+        if (pos >= end)
+            return DECODE_ABORT_LENGTH;
+
+        if (bitrow_get_bit(b->bb[row], pos++) != 0)
+            return DECODE_FAIL_SANITY;
+
+        if (pos + 8 >= end)
+            return DECODE_ABORT_LENGTH;
+
+        for (unsigned i = 0; i < 8; i++)
+            *out = *out << 1 | bitrow_get_bit(b->bb[row], pos++);
+
+        if (pos >= end)
+            return DECODE_ABORT_LENGTH;
+
+        if (bitrow_get_bit(b->bb[row], pos++) != 1)
+            return DECODE_FAIL_SANITY;
+
+        if (pos >= end)
+            return DECODE_ABORT_LENGTH;
+    }
+
+    return 10;
+}
+
+uint8_t next(const uint8_t* bb, unsigned *ipos, unsigned num_bytes) {
+    uint8_t r = bitrow_get_byte(bb, *ipos);
+    *ipos += 8;
+    if (*ipos >= num_bytes*8)
+       return DECODE_FAIL_SANITY;
+    return r;
+}
+
+typedef struct {
+    uint8_t header;
+    uint8_t num_device_ids : 2;
+    uint8_t device_id[4][3];
+    uint16_t command;
+    uint8_t payload_length;
+    uint8_t *payload;
+    uint8_t unparsed_length;
+    uint8_t *unparsed;
+    uint8_t crc;
+} message_t;
+
+typedef struct {
+    uint8_t unknown_0; /* always == 0? */
+    uint8_t unknown_1; /* direction? */
+    uint8_t second;
+    uint8_t minute;
+    uint8_t hour : 5;
+    uint8_t day_of_week : 3;
+    uint8_t day;
+    uint8_t month;
+    uint8_t year[2];
+} X313F;
+
+
+data_t *add_hex_string(data_t *data, const char *name, const uint8_t *buffer, size_t buffer_length)
+{
+    if (buffer && buffer_length > 0)  {
+        char tstr[(buffer_length * 2) + 1];
+        char *p = tstr;
+        for (unsigned i = 0; i < buffer_length; i++, p+=2)
+            sprintf(p, "%02x", buffer[i]);
+        *p = '\0';
+        data = data_append(data, name, "", DATA_STRING, tstr, NULL);
+    }
+    return data;
+}
+
+typedef struct { uint8_t t; const char s[4]; } dev_map_entry_t;
+
+static const dev_map_entry_t device_map[] = {
+    { .t = 01, .s = "CTL" }, /* Controller */
+    { .t = 02, .s = "UFH" }, /* Underfloor heating (HCC80, HCE80) */
+    { .t = 03, .s = " 30" }, /* HCW82?? */
+    { .t = 04, .s = "TRV" }, /* Thermostatic radiator valve (HR80, HR91, HR92) */
+    { .t = 07, .s = "DHW" }, /* DHW sensor (CS92) */
+    { .t = 10, .s = "OTB" }, /* OpenTherm bridge (R8810) */
+    { .t = 12, .s = "THm" }, /* Thermostat with setpoint schedule control (DTS92E, CME921) */
+    { .t = 13, .s = "BDR" }, /* Wireless relay box (BDR91) (HC60NG too?) */
+    { .t = 17, .s = " 17" }, /* Dunno - Outside weather sensor? */
+    { .t = 18, .s = "HGI" }, /* Honeywell Gateway Interface (HGI80, HGS80) */
+    { .t = 22, .s = "THM" }, /* Thermostat with setpoint schedule control (DTS92E) */
+    { .t = 30, .s = "GWY" }, /* Gateway (e.g. RFG100?) */
+    { .t = 32, .s = "VNT" }, /* (HCE80) Ventilation (Nuaire VMS-23HB33, VMN-23LMH23) */
+    { .t = 34, .s = "STA" }, /* Thermostat (T87RF) */
+    { .t = 63, .s = "NUL" }, /* No device */
+};
```

Comment:
json is quite talkative. All strings in the comments should be enabled in the json output.

Reply:
It's unused code, `decode_device_id()` isn't called as `style` is fixed to `1`.
Is it four different IDs per packet or is it a four-part ID? Wouldn't the type from this table be really useful?

Reply:
I've never used the json output, but I can take a look.

It's between 1 and 3 device IDs per packet (don't ask me why the did that...) and the first 6 bit of each ID indicate the device type. I don't like the "type:serial" format (`style==0`) , but some people seem to prefer it. Probably because it's the default output format of the Honeywell HGI80 gateway, which seems to have been popular at some time.